### DISCUSSION TITLE
Multithreading

### DISCUSSION THREAD

**Discussion by pezalot** [paullouisageneau/libdatachannel#1048] [General]
Hello and 10x for your work I appreciate it!!
I have a question- Send/Receive via datachannel are they in separate threads? Because in libwebrtc are. If I have about 5000 peers  (sending and receiving messages with diff burst and rate at the same time) with diff DC and PC can libdatachannel manage properly this situation? If not tell me some workarounds.
10x

**Comment by pezalot** [comment]
Oh, I forget to say that time delay between packets is verry importent.

**Comment by paullouisageneau** [comment]
Thank you for your interest in libdatachannel. Most things are run in separate threads, including send and receive, so sending and receiving bursts of message at the same time shouldn't be an issue. The library has an internal thread pool, and doesn't spawn new threads for each peer connection, so it should be able to scale to a high number of connections.

**Reply by BartolomeyKant** [reply]
Hello! You didn't mention here, is it save to send from multiple threads simultaneously?

**Reply by paullouisageneau** [reply]
@BartolomeyKant Yes, it's safe.
